<div id="container" style="position:relative;">
<div style="float:left"><h1> Kinema </h1></div>
<div style="position:relative; float:right"><img style="height:140px" src ="Images/IMG_0458.jpg" />
</div>
</div>

## A Film recommendation system based on below the line department heads
<br>

**These departments and position include**:
<br>

- **Writing** - Writer, Publisher, Screenplay, Story By...
- **Camera** - Director of Photography
- **Electric** - Lighting
- **Grip** - Stunts, Rigging, SFX
- **Art** - Production Designer, Prop Master, Art Director, Scenic
- **GLAM** - Hair, Make Up, Wardrobe
- **Editing** - Editor(s)
- **Sound** - Scorer, Music Supervisor, Foley Artist
- **Color** - Colorist, SFX, Lighting Technician, 
- **Locations** - Location Manager

As someone who has been passionately Directing, Writing, Acting, and Working professionaly in the Film industry for the past 8 years in Los Angeles, CA there is one thing (it's actually many!) that must be said when it comes to the streaming of Film and Television. The standard film studios use to recommendation films (recommender systems) needs to be reimagined and implemented.
<br>

That's where the unsung heroes of film magic come in. The professionals, the dreamers, the artists that make films come together are overlooked and under represented within the film industry. Most film recommendation systems follow a Film Director, Actor, and Movie Rating approach while also splashing in some random films to give user's that "organic" experience giving them the facade of natural selection. There is something critically missing from this entire approach. 
<br>

***What makes the film?***
<br>

> *It is not what we see, it is what we feel.*



# Credits - Cleaning

#### Credits Dataset

* `cast`: A stringified list of dictionaries consisting of cast names and the corresponding characters they played.
* `crew`: A stringified list of dictionaries consisting of crew names and the function they performed.
* `id`: The TMDB ID of movie.

This notebook will be used solely to clean the credits.csv from IMDB. This will serve as the heart of the recommendation system as it is my belief that it is the crew that really makes a film come to life. 

In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import norm
from scipy import stats
from ast import literal_eval

import warnings; warnings.simplefilter('ignore')

Let's take a look at what is within the credits dataframe.

In [14]:
df = pd.read_csv('data/movies_metadata.csv')
df.head().transpose()

0  \
adult                                                              False   
belongs_to_collection  {'id': 10194, 'name': 'Toy Story Collection', ...   
budget                                                          30000000   
genres                 [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
homepage                            http://toystory.disney.com/toy-story   
id                                                                   862   
imdb_id                                                        tt0114709   
original_language                                                     en   
original_title                                                 Toy Story   
overview               Led by Woody, Andy's toys live happily in his ...   
popularity                                                       21.9469   
poster_path                             /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
production_companies      [{'name': 'Pixar Animation Studios', 'id': 3}]   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-10-30   
revenue                                                      3.73554e+08   
runtime                                                               81   
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]   
status                                                          Released   
tagline                                                              NaN   
title                                                          Toy Story   
video                                                              False   
vote_average                                                         7.7   
vote_count                                                          5415   

                                                                       1  \
adult                                                              False   
belongs_to_collection                                                NaN   
budget                                                          65000000   
genres                 [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
homepage                                                             NaN   
id                                                                  8844   
imdb_id                                                        tt0113497   
original_language                                                     en   
original_title                                                   Jumanji   
overview               When siblings Judy and Peter discover an encha...   
popularity                                                       17.0155   
poster_path                             /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
production_companies   [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-12-15   
revenue                                                      2.62797e+08   
runtime                                                              104   
spoken_languages       [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
status                                                          Released   
tagline                        Roll the dice and unleash the excitement!   
title                                                            Jumanji   
video                                                              False   
vote_average                                                         6.9   
vote_count                                                          2413   

                                                                       2  \
adult                                                              False   
belongs_to_collection  {'id': 119050, 'name': 'Grumpy Old Men Collect...   
budget                                   

In [15]:
credits_df = pd.read_csv('data/credits.csv')
credits_df

cast  \
0      [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3      [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4      [{'cast_id': 1, 'character': 'George Banks', '...   
...                                                  ...   
45471  [{'cast_id': 0, 'character': '', 'credit_id': ...   
45472  [{'cast_id': 1002, 'character': 'Sister Angela...   
45473  [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...   
45474  [{'cast_id': 2, 'character': '', 'credit_id': ...   
45475                                                 []   

                                                    crew      id  
0      [{'credit_id': '52fe4284c3a36847f8024f49', 'de...     862  
1      [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...    8844  
2      [{'credit_id': '52fe466a9251416c75077a89', 'de...   15602  
3      [{'credit_id': '52fe44779251416c91011acb', 'de...   31357  
4      [{'credit_id': '52fe44959251416c75039ed7', 'de...   11862  
...                                                  ...     ...  
45471  [{'credit_id': '5894a97d925141426c00818c', 'de...  439050  
45472  [{'credit_id': '52fe4af1c3a36847f81e9b15', 'de...  111109  
45473  [{'credit_id': '52fe4776c3a368484e0c8387', 'de...   67758  
45474  [{'credit_id': '533bccebc3a36844cf0011a7', 'de...  227506  
45475  [{'credit_id': '593e676c92514105b702e68e', 'de...  461257  

[45476 rows x 3 columns]

I will go ahead and drop the `cast` feature from the data frame and go from there.

In [16]:
# Create dataframe with certain features
crew_df = credits_df[['crew','id']]
crew_df

crew      id
0      [{'credit_id': '52fe4284c3a36847f8024f49', 'de...     862
1      [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...    8844
2      [{'credit_id': '52fe466a9251416c75077a89', 'de...   15602
3      [{'credit_id': '52fe44779251416c91011acb', 'de...   31357
4      [{'credit_id': '52fe44959251416c75039ed7', 'de...   11862
...                                                  ...     ...
45471  [{'credit_id': '5894a97d925141426c00818c', 'de...  439050
45472  [{'credit_id': '52fe4af1c3a36847f81e9b15', 'de...  111109
45473  [{'credit_id': '52fe4776c3a368484e0c8387', 'de...   67758
45474  [{'credit_id': '533bccebc3a36844cf0011a7', 'de...  227506
45475  [{'credit_id': '593e676c92514105b702e68e', 'de...  461257

[45476 rows x 2 columns]

In [17]:
print(crew_df.iloc[0]['crew'])

[{'credit_id': '52fe4284c3a36847f8024f49', 'department': 'Directing', 'gender': 2, 'id': 7879, 'job': 'Director', 'name': 'John Lasseter', 'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'}, {'credit_id': '52fe4284c3a36847f8024f4f', 'department': 'Writing', 'gender': 2, 'id': 12891, 'job': 'Screenplay', 'name': 'Joss Whedon', 'profile_path': '/dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg'}, {'credit_id': '52fe4284c3a36847f8024f55', 'department': 'Writing', 'gender': 2, 'id': 7, 'job': 'Screenplay', 'name': 'Andrew Stanton', 'profile_path': '/pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg'}, {'credit_id': '52fe4284c3a36847f8024f5b', 'department': 'Writing', 'gender': 2, 'id': 12892, 'job': 'Screenplay', 'name': 'Joel Cohen', 'profile_path': '/dAubAiZcvKFbboWlj7oXOkZnTSu.jpg'}, {'credit_id': '52fe4284c3a36847f8024f61', 'department': 'Writing', 'gender': 0, 'id': 12893, 'job': 'Screenplay', 'name': 'Alec Sokolow', 'profile_path': '/v79vlRYi94BZUQnkkyznbGUZLjT.jpg'}, {'credit_id': '52fe4284c3a36847f8024f67', 'depart

So from the looks of it the data can be easily identified. For example let's take a look at the first line of data - and let's break it down into a more readable format.

<br> 

`{\'credit_id\': \'52fe4284c3a36847f8024f49\', \'department\': \'Directing\', \'gender\': 2, \'id\': 7879, \'job\': \'Director\', \'name\': \'John Lasseter\', \'profile_path\': \'/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg\'}`

<br>

- `credit_id` = **52fe4284c3a36847f8024f49** 
- `department` = **Directing** 
- `gender` = **2**
- `id` = **7879**
- `job` = **Director**
- `name` = **John Lasseter**
- `profile_path` = **7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg**

<br>

From here we can create a new dataframe for easy access and smooth analysis.




In [18]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [19]:
df['id'] = df['id'].apply(convert_int)

In [20]:
df[df['id'].isnull()]

adult  \
19730                                 - Written by Ørnås   
29503   Rune Balot goes to a casino connected to the ...   
35587   Avalanche Sharks tells the story of a bikini ...   

      belongs_to_collection                            budget  \
19730              0.065736  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29503              1.931659  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35587              2.185485  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage  id imdb_id  \
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso... NaN       0   
29503  [{'iso_3166_1': 'US', 'name': 'United States o... NaN       0   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}] NaN       0   

      original_language                            original_title  overview  \
19730             104.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
29503              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]  Released   
35587              82.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

       ... release_date revenue runtime spoken_languages status  tagline  \
19730  ...            1     NaN     NaN              NaN    NaN      NaN   
29503  ...           12     NaN     NaN              NaN    NaN      NaN   
35587  ...           22     NaN     NaN              NaN    NaN      NaN   

       title video vote_average vote_count  
19730    NaN   NaN          NaN        NaN  
29503    NaN   NaN          NaN        NaN  
35587    NaN   NaN          NaN        NaN  

[3 rows x 24 columns]

## Metadata Cleaning - (Crew)

In [28]:
credits = pd.read_csv('data/credits.csv')
keywords = pd.read_csv('data/keywords.csv')
metadata = pd.read_csv('data/movies_metadata.csv')

In [29]:
metadata['genres'] = metadata['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [30]:
metadata['year'] = pd.to_datetime(metadata['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [31]:
metadata = metadata.drop([19730, 29503, 35587])

In [32]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
metadata['id'] = metadata['id'].astype('int')

In [33]:
metadata.shape

(45463, 25)

In [34]:
metadata = metadata.merge(credits, on='id')
metadata = metadata.merge(keywords, on='id')

In [35]:
metadata.head(2)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   

                         genres                              homepage    id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story   862   
1  [Adventure, Fantasy, Family]                                   NaN  8844   

     imdb_id original_language original_title  \
0  tt0114709                en      Toy Story   
1  tt0113497                en        Jumanji   

                                            overview  ...    status  \
0  Led by Woody, Andy's toys live happily in his ...  ...  Released   
1  When siblings Judy and Peter discover an encha...  ...  Released   

                                     tagline      title  video vote_average  \
0                                        NaN  Toy Story  False          7.7   
1  Roll the dice and unleash the excitement!    Jumanji  False          6.9   

   vote_count  year                                               cast  \
0      5415.0  1995  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      2413.0  1995  [{'cast_id': 1, 'character': 'Alan Parrish', '...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  

[2 rows x 28 columns]

In [36]:
links_small = pd.read_csv('data/links_small.csv')
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')

In [37]:
crew_md = metadata[metadata['id'].isin(links_small)]
crew_md.shape

(9219, 28)

In [38]:
crew_md.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   

     imdb_id original_language    original_title  \
0  tt0114709                en         Toy Story   
1  tt0113497                en           Jumanji   
2  tt0113228                en  Grumpier Old Men   

                                            overview  ...    status  \
0  Led by Woody, Andy's toys live happily in his ...  ...  Released   
1  When siblings Judy and Peter discover an encha...  ...  Released   
2  A family wedding reignites the ancient feud be...  ...  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average  vote_count  year  \
0          7.7      5415.0  1995   
1          6.9      2413.0  1995   
2          6.5        92.0  1995   

                                                cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   

                                                crew  \
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...   
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...   

                                            keywords  
0  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...  
1  [{'id': 10090, 'name': 'board game'}, {'id': 1...  
2  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...  

[3 rows x 28 columns]

**CREW**: Most recommendation systems that I have come across disregard the crew and only keep the Director for their system - thinking that the director is the sole proprieter of the "Feel" of a film. This is wrong, and I cannot stress enough that a film director highly relies on the artisanship and expertise of many of those that make up a film crew. Such as:

<br>

- Cinematographer
- Production Designer
- Lighting
- Hair / Make Up / Wardrobe
- Editor
- Colorist

and that is to name just a few. Film directors do not make movies alone, filmmaking is a highly collaborative effort and each department leans on one another. Directors, the good directors are those that can be humble and take the advice from the departments heads on and off set.

<br>

Take for instance - "Jaws" - Directed by Steven Spielberg. Was it not for the great Verna Fields who had been a long time editor working with the likes of Alfred Hitchcock, there would be no "Jaws" today. She understood and learned from the great suspense director Alfred Hitchcock that if you only enveil the "Horror" for a split second and let the viewer's mind do the rest - that it can actually be even more terrorfying than showing the "Horror" completely. This is was Verna taught Steven the day he returned to Universal Studios after a disastrous shoot day in which they could not get the mechainal shark to work. The unsung heroes like Verna are an incredibly important part to filmmaking as they bring the combined magic of those that came before them.

In [47]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [48]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [49]:
smd['director'] = smd['crew'].apply(get_director)

In [51]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [52]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

In [53]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [54]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64